connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -366,7 +367,11 @@ private boolean sendRecords() {
                         if (e != null) {
                             log.error("{} failed to send record to {}: ", WorkerSourceTask.this, topic, e);
```

Comment:
Should we modify this line to respect the [errors.log.enable](https://kafka.apache.org/30/documentation.html#sourceconnectorconfigs_errors.log.enable) (and possibly [errors.log.include.messages](https://kafka.apache.org/30/documentation.html#sourceconnectorconfigs_errors.log.include.messages)) properties?

I wonder if it might be useful to still unconditionally set `producerSendException` (or perhaps even convert that field from an `AtomicReference<Throwable>` to some kind of list, and append to it here) and then modify the contents (and possibly also name) of `maybeThrowProducerSendException` to have our error-handling logic. Thoughts?

Reply:
Actually, that may complicate things by causing records to be given to `SourceTask::commitRecord` out of order (a record that caused a producer failure may be committed after a record that was dispatched to the producer after it). So probably best to keep the error-handling logic here, but I do still wonder if we can respect the logging-related configuration properties.

Reply:
Now that this could be a tolerated error, it makes sense to have it respect the errors.log.enable configuration, but the log line would be duplicated, unconditionally writing it in the event we do not tolerate and a config check if we do. 

Are you envisioning something like this?

```
if (retryWithToleranceOperator.getErrorToleranceType().equals(ToleranceType.ALL)) {
    if (errorLogEnabled) { // get this value from the config in some manner
        log.error("{} failed to send record to {}: ", WorkerSourceTask.this, topic, e);
        log.trace("{} Failed record: {}", WorkerSourceTask.this, preTransformRecord);
    }
    commitTaskRecord(preTransformRecord, null);
} else {
    log.error("{} failed to send record to {}: ", WorkerSourceTask.this, topic, e);
    log.trace("{} Failed record: {}", WorkerSourceTask.this, preTransformRecord);
    producerSendException.compareAndSet(null, e);
}
```

I would need to look more closely at the other layers of objects on top of the SourceTask. enableErrorLog() is available in the ConnectorConfig, but only the SinkConnectorConfig makes use of it. I would need to spin up some additional infrastructure. Not sure if I would want to add WorkerErrantRecordReporter to WorkerSourceTask or have the configuration pass down in some other manner.

Reply:
Yes, I was thinking the behavior could be something like that code snippet, although we'd also want to respect the [errors.log.include.messages](https://kafka.apache.org/30/documentation.html#sourceconnectorconfigs_errors.log.include.messages) property and would probably want the format of the error messages to be similar to the error messages we emit in other places where messages are tolerated (such as when conversion or transformation fails).

Reply:
The error retry handling infrastructure predominantly concerns itself with the sink side of the house. Insofar that any refactoring I would want to do would probably necessitate a KIP on its own. To that end, I have added an additional executeFailed() function to RetryWithToleranceOperator to allow the source worker to handle error logging with all of the existing infrastructure/configuration that exists for sink tasks.

I toy'ed around with the idea of having the new executeFailed() fire without a tolerance type check. This would work for failing/ignoring as expected, but with no mechanism to then decide if we should call commitRecord(). We could block on the future from executeFailed() and then check withinToleranceLimits() but that introduces non determinism with interrupt/execution exceptions.